### PR TITLE
hook test for twap

### PR DESCRIPTION
**Pull request by catShaark**
Closes: #2211

## What is the purpose of the change

> Add a description of the overall background and high level changes that this PR introduces

*(E.g.: This pull request improves documation of area A by adding ....*

## Brief Changelog

*(for example:)*
 
  - *The metadata is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *Daemons retrieve the RPC data from the blob cache*

## Testing and Verifying

*(Please pick one of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added unit test that validates ...*
  - *Added integration tests for end-to-end deployment with ...*
  - *Extended integration test for ...*
  - *Manually verified the change by ...*

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: app/apptesting/gamm.go [hunk 1]
```diff
@@ -114,3 +114,26 @@ func (s *KeeperTestHelper) RunBasicSwap(poolId uint64) {
...
+	denoms, err := s.App.GAMMKeeper.GetPoolDenoms(s.Ctx, poolId)
+	s.Require().NoError(err)
+
+	var outDenom string
+	if denoms[0] != tokenIn.Denom {
+		outDenom = denoms[0]
+	} else {
+		outDenom = denoms[1]
+	}
+	s.FundAcc(s.TestAccs[0], sdk.Coins{tokenIn})
```
Comment (alexanderbez) [inline app/apptesting/gamm.go:128]:
```suggestion
	s.FundAcc(s.TestAccs[0], sdk.NewCoins(tokenIn))
```

File: app/apptesting/gamm.go [hunk 2]
```diff
@@ -114,3 +114,26 @@ func (s *KeeperTestHelper) RunBasicSwap(poolId uint64) {
...
+
+	msg := gammtypes.MsgSwapExactAmountIn{
+		Sender:            string(s.TestAccs[0]),
+		Routes:            []gammtypes.SwapAmountInRoute{{PoolId: poolId, TokenOutDenom: outDenom}},
+		TokenIn:           tokenIn,
+		TokenOutMinAmount: sdk.ZeroInt(),
+	}
+	// TODO: switch to message
```
Comment (alexanderbez) [inline app/apptesting/gamm.go:136]: Is this TODO still relevant? Can we reference an issue if it's not?

File: x/gamm/twap/hook_test.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+	"github.com/osmosis-labs/osmosis/v10/x/gamm/twap/types"
+)
+
+// TestSwapAndEndBlockTriggeringSave tests that if we:
+// * create a pool in block 1
+// * swap in block 2
+// then after block 2 end block, we have saved records for the pool,
+// for both block 1 & 2, with distinct spot prices in their records, and accumulators incremented.
+// TODO: Abstract this to be more table driven, and test more pool / block setups.
```
Comment (alexanderbez) [inline x/gamm/twap/hook_test.go:16]: ditto

File: app/apptesting/gamm.go (hunk 2 above)
Comment (catShaark) [reply app/apptesting/gamm.go:136]: I'll address this in this PR

File: app/apptesting/gamm.go [hunk 4]
```diff
@@ -110,7 +111,32 @@ func (s *KeeperTestHelper) RunBasicSwap(poolId uint64) {
...
 	}
-	// TODO: switch to message
-	_, err = s.App.GAMMKeeper.SwapExactAmountIn(s.Ctx, s.TestAccs[0], poolId, msg.TokenIn, denoms[1], msg.TokenOutMinAmount)
+
+	gammMsgServer := gammkeeper.NewMsgServerImpl(s.App.GAMMKeeper)
+	_, err = gammMsgServer.SwapExactAmountIn(sdk.WrapSDKContext(s.Ctx), &msg)
+	s.Require().NoError(err)
+}
+
+func (s *KeeperTestHelper) RunCustomSwap(poolId uint64, tokenIn sdk.Coin) {
```
[the remark this answers is not in the document]
Comment (mattverse) [reply app/apptesting/gamm.go:120]: and then we can replace the whole if doswap field and swap height field in the test case below

File: x/gamm/twap/hook_test.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+		name string
+
+		// this pool will be created at height 1
+		createPoolFunc func() uint64
+
+		// token to be swaped in at each block height
+		tokenInAtBlock map[int64]sdk.Coin
+
+		// block height to stop at
+		stopHeight int64
```
Comment (mattverse) [inline x/gamm/twap/hook_test.go:28]:
```suggestion
	tests := []struct {
		name string
		// this pool will be created at height 1
		createPoolFunc func() uint64
		// token to be swaped in at each block height
		tokenInAtBlock map[int64]sdk.Coin
		// block height to stop at
		stopHeight int64
```

File: x/gamm/twap/hook_test.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
+package twap_test
```
[the remark this answers is not in the document]
Comment (catShaark) [reply x/gamm/twap/hook_test.go:1]: yess

File: app/apptesting/gamm.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (catShaark) [reply app/apptesting/gamm.go:120]: I think this function might be used for other tests that don't care about height

File: x/gamm/twap/hook_test.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	for _, tc := range tests {
+		s.Run(tc.name, func() {
+			s.SetupTest()
+
+			poolId := tc.createPoolFunc()
+			twapAtPoolCreation, err := types.NewTwapRecord(s.App.GAMMKeeper, s.Ctx, poolId, denom0, denom1)
+			s.Require().NoError(err)
+			expectedHistoricalTwap := twapAtPoolCreation
```
[the remark this answers is not in the document]
Comment (catShaark) [reply x/gamm/twap/hook_test.go]: what should I rename this to ?

[the remark this answers is not in the document]
Comment (mattverse) [reply x/gamm/twap/hook_test.go]: I think a one liner documentation / comment would suffice

File: app/apptesting/gamm.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattverse) [reply app/apptesting/gamm.go:120]: Hrmmm I feel like we can reduce so much code if we have a apptesting method that takes in height and swap, (maybe a function we want to run before end blocker) as parameters and then have it run swap + run given function + end block + commit in the designated hieght. Also could be useful in many other hooks test or end blockers tests(superfluid tests for example). 

We can come back to this though, not an important thing that should block merging this PR

**Comment by catShaark** [comment]
since twap has moved to its own folder, I'll close this PR and make a new one

**Pull request closed without merging**